### ISSUE TITLE
Radio FM

### ISSUE THREAD

**Issue by stucki** [aquaris-dev/DEPRECATED_android_device_bq_piccolo#87]
_From @angelhp on September 13, 2015 1:4_

_Copied from original issue: piccolo-dev/android_device_bq_piccolo#1_

**Comment by stucki** [comment]
_From @Kra1o5 on September 13, 2015 1:29_

Camera FM?

What is this?

**Comment by stucki** [comment]
_From @angelhp on September 14, 2015 7:24_

Sorry , radio fm not found

**Comment by stucki** [comment]
_From @angelhp on September 14, 2015 7:26_

The batery not arrive to 100%
Is in 96%

**Comment by stucki** [comment]
_From @Kra1o5 on September 14, 2015 7:59_

Not found or not working correctly?

If is the second, is a know issue.

For battery is really weird I get 100% without problems.

**Comment by stucki** [comment]
Hey @angelhp! Please create 1 ticket per issue. This one is about Camera FM. The status is, it works for 3 seconds only. I had in mind that @Kra1o5 already fixed this but it still seems to exist.

**Comment by stucki** [comment]
_From @Kra1o5 on September 14, 2015 8:25_

Yes @stucki .

But really is Radio FM not Camera is a typo error.

We have located the error on audio HAL & we are looking one method to fix it.

**Comment by stucki** [comment]
Right. Can you fix the title of this ticket?

**Comment by stucki** [comment]
_From @Kra1o5 on September 14, 2015 8:48_

@stucki Fixed!

**Comment by stucki** [comment]
_From @SIRSteiner on September 18, 2015 20:21_

Hello! FM-Radio doesn't crash but there is no output. Recording works. Thanks!

**Comment by stucki** [comment]
_From @Kra1o5 on September 18, 2015 20:22_

Yes, we have located the problem (audio HAL).

But not fixed yet.

**Comment by stucki** [comment]
_From @andy-de on September 27, 2015 15:42_

Spirit 2 works fine. Only switching output from headphone to internal speaker does not work. Maybe it helps solving this issues.

**Comment by stucki** [comment]
_From @Kra1o5 on September 28, 2015 11:7_

Hi @andy-de ,

We have the issue located but thanks for the tip, people can use Spirit 2 as temporal workaround.

Greetings

**Comment by stucki** [comment]
_From @Kra1o5 on October 31, 2015 22:37_

By request of @stucki ,

More info about the issue.

Is located in audio.primary.msm8916.so (/system/lib/hw/).

It seems like stock lib make something on it to avoid cut audio on kernel suspend.

We have two options.

Fix it on kernel (sounds part) or try to fix it on audio HAL part (audio.primary.msm8916.so).

**Comment by stucki** [comment]
_From @starvoid on December 14, 2015 9:31_

Hi, the fix for this issue....

could be get from this kernel?

https://github.com/CyanogenMod/android_kernel_bq_msm8916

I'm not sure, but I guess that is for aquaris X5. 

The last change was 3 days ago. It seems an active develop

Thanks for your effort.

**Comment by stucki** [comment]
_From @piwats on December 28, 2015 22:31_

Does the FM Radio app from the bq firmware run on cm?

**Comment by stucki** [comment]
_From @Kra1o5 on December 28, 2015 22:36_

The problem is on audio HAL.

**Comment by stucki** [comment]
_From @piwats on December 28, 2015 23:20_

Please excuse me, if I'm getting annoying.

I was looking the reference cynogenmod msm8939 device description files and noticed that they have a slightly different uevent.qcom.d file, where you can find the following line:

https://github.com/CyanogenMod/android_device_cyanogen_msm8939-common/blob/72c919fe35521ae0b907c1eb755ddd7c17c4e021/rootdir/etc/ueventd.qcom.rc#L138

```
/dev/msm_fm               0660   system     audio
```

via: https://github.com/CyanogenMod/android_device_cyanogen_msm8939-common/search?utf8=%E2%9C%93&q=fm

A quick search over at codeaurora (caf, msm sources or how else they are called) revealed, that they do something like this (via https://www.codeaurora.org/cgit/external/gigabyte/ag-gb-dsds-7227/plain/hardware/msm7k/libaudio-msm8660/AudioHardware.cpp)

```
#define FM_DEVICE  "/dev/msm_fm"
```

Do you know where "our" /dev/msm_fm has gone??

UPDATE: found a question to this, this one is obsolete, as radio from the kernel perspective is working, spirit radio app proves it.

**Comment by stucki** [comment]
_From @piwats on December 31, 2015 3:7_

Looking at the repo for the Xiami ferrari (msm8939) - it seems they have a working fm radio -, I noticed the following:

There is a

```
init.qcom.fm.sh
```

Look here: https://github.com/dh-harald/android_device_xiaomi_ferrari/blob/cm-12.1/rootdir/etc/init.qcom.fm.sh

which makes use of the following files/blobs/binaries included also in the original bq firmware (init.qcom.rc.sh is also present in the bq firmware):

```
# FM
bin/fmconfig
bin/fmfactorytest
bin/fmfactorytestserver
bin/fm_qsoc_patches
bin/ftmdaemon
```

It is worth mentioning, that google is exploding when trying to search for "init.qcom.fm.sh" as it seems to "turn on" fm radio where there was none before.

Happy new year!

**Comment by stucki** [comment]
_From @piwats on January 20, 2016 23:13_

Seems to work in CM-13 (cm-13-20160120) :+1:

**Comment by stucki** [comment]
Issue seems to be fixed in cm-13.0, therefore I'm closing this issue.
